### ISSUE TITLE
/usr/bin/ld: warning: libssl.so.10, needed by ./.libs/libevent_openssl.so, may conflict with libssl.so.3

### ISSUE THREAD

**Issue by mincymathew** [libevent/libevent#1142]
I have upgraded to openssl 3.0 on my system from openssl 1.0.2 and I am trying to make libevent. I used the following commands.

 ./configure LDFLAGS="-L/usr/src/openssl/" CFLAGS="-I/usr/src/openssl/include"
make

=========
Output 
=========
  CC       buffer.lo
  CCLD     libevent.la
  CCLD     sample/dns-example
  CCLD     sample/event-read-fifo
  CCLD     sample/hello-world
  CCLD     sample/http-server
  CCLD     sample/http-connect
  CCLD     sample/signal-test
  CCLD     sample/time-test
  CCLD     sample/le-proxy
/usr/bin/ld: warning: libssl.so.10, needed by ./.libs/libevent_openssl.so, may conflict with libssl.so.3
/usr/bin/ld: warning: libcrypto.so.10, needed by ./.libs/libevent_openssl.so, may conflict with libcrypto.so.3
  CCLD     sample/https-client
/usr/bin/ld: warning: libssl.so.10, needed by ./.libs/libevent_openssl.so, may conflict with libssl.so.3
/usr/bin/ld: warning: libcrypto.so.10, needed by ./.libs/libevent_openssl.so, may conflict with libcrypto.so.3
  CCLD     test/bench
  CCLD     test/bench_cascade
  CCLD     test/bench_http
  CCLD     libevent_core.la
  CCLD     test/bench_httpclient
  CCLD     test/test-changelist
  CCLD     test/test-dumpevents
  CCLD     test/test-eof
  CCLD     test/test-closed
  CCLD     test/test-fdleak
  CCLD     test/test-init
  CCLD     test/test-ratelim
  CCLD     test/test-time
  CCLD     test/test-weof
  CCLD     test/regress
/usr/bin/ld: warning: libssl.so.10, needed by ./.libs/libevent_openssl.so, may conflict with libssl.so.3
/usr/bin/ld: warning: libcrypto.so.10, needed by ./.libs/libevent_openssl.so, may conflict with libcrypto.so.3

How can I resolve this warning?

**Comment by ploxiln** [comment]
It appears that `libevent_openssl.so` was linked with openssl-1.0.2 while the examples and tests were linked with openssl-3.0, somehow

**Comment by ploxiln** [comment]
(I wonder if your source/build dir were not completely clean)

**Comment by mincymathew** [comment]
You are right. After cleaning the dir, it was built correctly. However, I don't see ktls getting used for the data IO. I have built openssl with enable-ktls. lsmod shows it. But tls_stat does not increment. Is there any specific flag I need use?

**Comment by ploxiln** [comment]
I don't think libevent offers any flags to enable that kind of feature. Your best bet when writing your own application is to enable it in the `SSL*` that you create separately and provide to libevent e.g. in [bufferevent_openssl_socket_new()](http://www.wangafu.net/~nickm/libevent-2.1/doxygen/html/bufferevent__ssl_8h.html#a7a3090b586bfc7fc02e81a3a2c2deea5)

**Comment by mincymathew** [comment]
On a regular TCP connection,  when a client hello is detected, instead of calling bufferevent_openssl_filter_new over the bev,  can the underlying bufferevent be closed and then bufferevent_openssl_socket_new be called, callbacks set and enabled?

**Comment by ploxiln** [comment]
`bufferevent_openssl_filter_new()` also takes a custom `SSL*` (aka `struct ssl_st*`)

**Comment by mincymathew** [comment]
I need to call the BIO_new_socket which has the ktls BIO methods in openssl 3.0.

**Comment by ploxiln** [comment]
I don't know how to do that with a libevent bufferevent, sorry.

**Comment by mincymathew** [comment]
bufferevent_openssl_socket_new calls BIO_new_socket. 
If I close a regular bufferevent of a tcp connection, instead of calling bufferevent_openssl_filter_new over it, and then call bufferevent_openssl_socket_new, the handshake isn't proceeding. Not sure why.

**Comment by mincymathew** [comment]
Why is do_handshake () called in set_handshake_callbacks () only for bufferevents with underlying bufferevents ?

**Comment by azat** [comment]
>You are right. After cleaning the dir, it was built correctly. However, I don't see ktls getting used for the data IO. I have built openssl with enable-ktls. lsmod shows it. But tls_stat does not increment. Is there any specific flag I need use?

libevent does not have support of ktls, but AFAICS you just need to set two options for the socket:

```c
/// fill crypto_info
assert(setsockopt(client, SOL_TCP, TCP_ULP, "tls", sizeof("tls")) == 0);
assert(setsockopt(client, SOL_TLS, TLS_TX, &crypto_info, sizeof(crypto_info)) == 0);
```

>On a regular TCP connection, when a client hello is detected, instead of calling bufferevent_openssl_filter_new over the bev, can the underlying bufferevent be closed and then bufferevent_openssl_socket_new be called, callbacks set and enabled?

What for?

>I need to call the BIO_new_socket which has the ktls BIO methods in openssl 3.0.

I [don't see this](https://www.openssl.org/docs/manmaster/man3/BIO_s_socket.html), can provide an example what does not work?

>bufferevent_openssl_socket_new calls BIO_new_socket.

Only if you not pass the fd, but you can create the fd and pass existing.

>If I close a regular bufferevent of a tcp connection, instead of calling bufferevent_openssl_filter_new over it, and then call bufferevent_openssl_socket_new, the handshake isn't proceeding. Not sure why.

Please provide a sample.

>Why is do_handshake () called in set_handshake_callbacks () only for bufferevents with underlying bufferevents ?

do_handshake should be called from be_ssl_handshakeeventcb in this case.
